File: agenkit-cpp/tests/unit/test_optimizer.cpp
/**
 * @file test_optimizer.cpp
 * @brief Unit tests for optimizer framework
 */

#include <gtest/gtest.h>
#include "agenkit/evaluation/optimizer.hpp"
#include <thread>
#include <chrono>

using namespace agenkit::evaluation;

// SearchSpace Tests

TEST(SearchSpaceTest, AddContinuousParameter) {
    SearchSpace space;
    space.add_continuous("temperature", 0.0, 1.0);

    const auto& params = space.parameters();
    ASSERT_EQ(params.size(), 1);
    ASSERT_EQ(params.count("temperature"), 1);
    EXPECT_EQ(params.at("temperature").type, ParameterType::Continuous);
    EXPECT_DOUBLE_EQ(params.at("temperature").low, 0.0);
    EXPECT_DOUBLE_EQ(params.at("temperature").high, 1.0);
}

TEST(SearchSpaceTest, AddDiscreteParameter) {
    SearchSpace space;
    space.add_discrete("max_tokens", {128.0, 256.0, 512.0});

    const auto& params = space.parameters();
    ASSERT_EQ(params.size(), 1);
    ASSERT_EQ(params.count("max_tokens"), 1);
    EXPECT_EQ(params.at("max_tokens").type, ParameterType::Discrete);
    EXPECT_EQ(params.at("max_tokens").values.size(), 3);
}

TEST(SearchSpaceTest, AddIntegerParameter) {
    SearchSpace space;
    space.add_integer("top_k", 1, 100);

    const auto& params = space.parameters();
    ASSERT_EQ(params.size(), 1);
    ASSERT_EQ(params.count("top_k"), 1);
    EXPECT_EQ(params.at("top_k").type, ParameterType::Integer);
    EXPECT_DOUBLE_EQ(params.at("top_k").low, 1.0);
    EXPECT_DOUBLE_EQ(params.at("top_k").high, 100.0);
}

TEST(SearchSpaceTest, AddCategoricalParameter) {
    SearchSpace space;
    space.add_categorical("model", {"gpt-4", "claude-3", "gemini"});

    const auto& params = space.parameters();
    ASSERT_EQ(params.size(), 1);
    ASSERT_EQ(params.count("model"), 1);
    EXPECT_EQ(params.at("model").type, ParameterType::Categorical);
    EXPECT_EQ(params.at("model").values.size(), 3);
}

TEST(SearchSpaceTest, SampleContinuous) {
    SearchSpace space;
    space.add_continuous("temperature", 0.0, 1.0);

    // Sample multiple times to test distribution
    for (int i = 0; i < 10; ++i) {
        auto config = space.sample();
        ASSERT_EQ(config.size(), 1);
        ASSERT_EQ(config.count("temperature"), 1);

        double value = std::any_cast<double>(config.at("temperature"));
        EXPECT_GE(value, 0.0);
        EXPECT_LE(value, 1.0);
    }
}

TEST(SearchSpaceTest, SampleInteger) {
    SearchSpace space;
    space.add_integer("top_k", 1, 10);

    // Sample multiple times
    for (int i = 0; i < 10; ++i) {
        auto config = space.sample();
        ASSERT_EQ(config.size(), 1);
        ASSERT_EQ(config.count("top_k"), 1);

        int value = std::any_cast<int>(config.at("top_k"));
        EXPECT_GE(value, 1);
        EXPECT_LE(value, 10);
    }
}

TEST(SearchSpaceTest, SampleDiscrete) {
    SearchSpace space;
    space.add_discrete("max_tokens", {128.0, 256.0, 512.0});

    std::set<double> seen_values;
    for (int i = 0; i < 20; ++i) {
        auto config = space.sample();
        double value = std::any_cast<double>(config.at("max_tokens"));
        seen_values.insert(value);

        EXPECT_TRUE(value == 128.0 || value == 256.0 || value == 512.0);
    }

    // Should see at least 2 different values in 20 samples (probabilistically)
    EXPECT_GE(seen_values.size(), 2);
}

TEST(SearchSpaceTest, SampleCategorical) {
    SearchSpace space;
    space.add_categorical("model", {"gpt-4", "claude-3"});

    std::set<std::string> seen_values;
    for (int i = 0; i < 20; ++i) {
        auto config = space.sample();
        std::string value = std::any_cast<std::string>(config.at("model"));
        seen_values.insert(value);

        EXPECT_TRUE(value == "gpt-4" || value == "claude-3");
    }

    // Should see both values in 20 samples (probabilistically)
    EXPECT_EQ(seen_values.size(), 2);
}

TEST(SearchSpaceTest, SampleMultipleParameters) {
    SearchSpace space;
    space.add_continuous("temperature", 0.0, 1.0);
    space.add_integer("top_k", 1, 100);
    space.add_categorical("model", {"gpt-4", "claude-3"});

    auto config = space.sample();
    EXPECT_EQ(config.size(), 3);
    EXPECT_EQ(config.count("temperature"), 1);
    EXPECT_EQ(config.count("top_k"), 1);
    EXPECT_EQ(config.count("model"), 1);
}

TEST(SearchSpaceTest, ValidateCorrectConfig) {
    SearchSpace space;
    space.add_continuous("temperature", 0.0, 1.0);
    space.add_integer("top_k", 1, 100);

    std::map<std::string, std::any> config;
    config["temperature"] = 0.7;
    config["top_k"] = 50;

    EXPECT_TRUE(space.validate(config));
}

TEST(SearchSpaceTest, ValidateOutOfRangeContinuous) {
    SearchSpace space;
    space.add_continuous("temperature", 0.0, 1.0);

    std::map<std::string, std::any> config;
    config["temperature"] = 1.5;  // Out of range

    EXPECT_FALSE(space.validate(config));
}

TEST(SearchSpaceTest, ValidateOutOfRangeInteger) {
    SearchSpace space;
    space.add_integer("top_k", 1, 100);

    std::map<std::string, std::any> config;
    config["top_k"] = 150;  // Out of range

    EXPECT_FALSE(space.validate(config));
}

TEST(SearchSpaceTest, ValidateMissingParameter) {
    SearchSpace space;
    space.add_continuous("temperature", 0.0, 1.0);
    space.add_integer("top_k", 1, 100);

    std::map<std::string, std::any> config;
    config["temperature"] = 0.7;  // Missing top_k

    EXPECT_FALSE(space.validate(config));
}

TEST(SearchSpaceTest, ValidateUnknownParameter) {
    SearchSpace space;
    space.add_continuous("temperature", 0.0, 1.0);

    std::map<std::string, std::any> config;
    config["temperature"] = 0.7;
    config["unknown_param"] = 42;  // Unknown parameter

    EXPECT_FALSE(space.validate(config));
}

TEST(SearchSpaceTest, ValidateInvalidDiscreteValue) {
    SearchSpace space;
    space.add_discrete("max_tokens", {128.0, 256.0, 512.0});

    std::map<std::string, std::any> config;
    config["max_tokens"] = 300.0;  // Not in list

    EXPECT_FALSE(space.validate(config));
}

TEST(SearchSpaceTest, ValidateInvalidCategoricalValue) {
    SearchSpace space;
    space.add_categorical("model", {"gpt-4", "claude-3"});

    std::map<std::string, std::any> config;
    config["model"] = std::string("unknown-model");  // Not in list

    EXPECT_FALSE(space.validate(config));
}

// OptimizationResult Tests

TEST(OptimizationResultTest, DurationSeconds) {
    OptimizationResult result;
    result.start_time = std::chrono::system_clock::now();
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    result.end_time = std::chrono::system_clock::now();

    double duration = result.duration_seconds();
    EXPECT_GE(duration, 0.1);
    EXPECT_LT(duration, 0.2);  // Should be close to 0.1s
}

TEST(OptimizationResultTest, GetImprovement) {
    OptimizationResult result;

    OptimizationStep step1;
    step1.score = 0.5;
    result.history.push_back(step1);

    OptimizationStep step2;
    step2.score = 0.7;
    result.history.push_back(step2);

    result.best_score = 0.7;

    double improvement = result.get_improvement();
    EXPECT_DOUBLE_EQ(improvement, 40.0);  // (0.7 - 0.5) / 0.5 * 100 = 40%
}

TEST(OptimizationResultTest, GetImprovementEmptyHistory) {
    OptimizationResult result;
    double improvement = result.get_improvement();
    EXPECT_DOUBLE_EQ(improvement, 0.0);
}

TEST(OptimizationResultTest, ToJsonFromJson) {
    OptimizationResult result;
    result.best_config["temperature"] = 0.7;
    result.best_config["top_k"] = 50;
    result.best_score = 0.85;
    result.n_iterations = 10;
    result.metadata["algorithm"] = "random_search";

    OptimizationStep step;
    step.config["temperature"] = 0.5;
    step.score = 0.75;
    result.history.push_back(step);

    // Serialize
    auto json = result.to_json();

    // Deserialize
    auto result2 = OptimizationResult::from_json(json);

    EXPECT_DOUBLE_EQ(result2.best_score, 0.85);
    EXPECT_EQ(result2.n_iterations, 10);
    EXPECT_EQ(result2.history.size(), 1);
    EXPECT_DOUBLE_EQ(result2.history[0].score, 0.75);
}

// RandomSearchOptimizer Tests

TEST(RandomSearchOptimizerTest, OptimizeMaximize) {
    // Create search space
    auto space = std::make_shared<SearchSpace>();
    space->add_continuous("x", -10.0, 10.0);

    // Objective: maximize -(x-3)^2 (peak at x=3)
    auto objective = [](const std::map<std::string, std::any>& config) -> double {
        double x = std::any_cast<double>(config.at("x"));
        return -(x - 3.0) * (x - 3.0);
    };

    // Run optimization
    RandomSearchOptimizer optimizer(objective, space, true);
    auto result = optimizer.optimize(50).get();

    EXPECT_EQ(result.n_iterations, 50);
    EXPECT_EQ(result.history.size(), 50);

    // Best x should be close to 3
    double best_x = std::any_cast<double>(result.best_config.at("x"));
    EXPECT_GE(best_x, 1.0);
    EXPECT_LE(best_x, 5.0);

    // Best score should be close to 0 (the maximum)
    EXPECT_GE(result.best_score, -4.0);  // Within reasonable range
}

TEST(RandomSearchOptimizerTest, OptimizeMinimize) {
    // Create search space
    auto space = std::make_shared<SearchSpace>();
    space->add_continuous("x", -10.0, 10.0);

    // Objective: minimize (x-3)^2 (minimum at x=3)
    auto objective = [](const std::map<std::string, std::any>& config) -> double {
        double x = std::any_cast<double>(config.at("x"));
        return (x - 3.0) * (x - 3.0);
    };

    // Run optimization (minimize)
    RandomSearchOptimizer optimizer(objective, space, false);
    auto result = optimizer.optimize(50).get();

    EXPECT_EQ(result.n_iterations, 50);

    // Best x should be close to 3
    double best_x = std::any_cast<double>(result.best_config.at("x"));
    EXPECT_GE(best_x, 1.0);
    EXPECT_LE(best_x, 5.0);

    // Best score should be negative (since we negate for minimization)
    // and close to 0 (the minimum)
    EXPECT_LE(result.best_score, 0.0);
    EXPECT_GE(result.best_score, -4.0);
}

TEST(RandomSearchOptimizerTest, OptimizeMultipleParameters) {
    // Create search space
    auto space = std::make_shared<SearchSpace>();
    space->add_continuous("x", -5.0, 5.0);
    space->add_continuous("y", -5.0, 5.0);

    // Objective: maximize -(x^2 + y^2) (peak at (0,0))
    auto objective = [](const std::map<std::string, std::any>& config) -> double {
        double x = std::any_cast<double>(config.at("x"));
        double y = std::any_cast<double>(config.at("y"));
        return -(x*x + y*y);
    };

    // Run optimization
    RandomSearchOptimizer optimizer(objective, space, true);
    auto result = optimizer.optimize(100).get();

    EXPECT_EQ(result.n_iterations, 100);

    // Best should be close to (0, 0)
    double best_x = std::any_cast<double>(result.best_config.at("x"));
    double best_y = std::any_cast<double>(result.best_config.at("y"));

    // With 100 samples, should get reasonably close
    EXPECT_GE(best_x, -3.0);
    EXPECT_LE(best_x, 3.0);
    EXPECT_GE(best_y, -3.0);
    EXPECT_LE(best_y, 3.0);

    // Best score should be close to 0
    EXPECT_GE(result.best_score, -10.0);
}

TEST(RandomSearchOptimizerTest, ImprovementCalculation) {
    auto space = std::make_shared<SearchSpace>();
    space->add_continuous("x", 0.0, 10.0);

    // Objective that improves
    auto objective = [](const std::map<std::string, std::any>& config) -> double {
        double x = std::any_cast<double>(config.at("x"));
        return x;  // Higher is better
    };

    RandomSearchOptimizer optimizer(objective, space, true);
    auto result = optimizer.optimize(20).get();

    // Should see improvement
    double improvement = result.get_improvement();
    EXPECT_GT(improvement, 0.0);
}

File: tests/evaluation/test_optimizer.py
"""
Tests for optimization framework base classes.
"""

from datetime import UTC

import pytest

from agenkit.evaluation import OptimizationResult, RandomSearchOptimizer, SearchSpace
from agenkit.interfaces import Message


class MockAgent:
    """Mock agent for testing."""

    def __init__(self, temperature: float = 0.7, top_p: float = 0.9):
        self.temperature = temperature
        self.top_p = top_p
        self.call_count = 0

    async def process(self, message: Message, session_id: str = "") -> Message:
        """Process message (mock implementation)."""
        self.call_count += 1
        # Simulate performance based on config
        # Better configs (lower temp, higher top_p) perform better
        score = (1 - self.temperature) * self.top_p
        is_correct = score > 0.5
        content = "correct answer" if is_correct else "incorrect"
        return Message(role="assistant", content=content)


# Test SearchSpace


def test_search_space_continuous():
    """Test continuous parameter definition."""
    space = SearchSpace()
    space.add_continuous("temperature", 0.0, 1.0)

    assert "temperature" in space.parameters
    assert space.parameters["temperature"]["type"] == "continuous"
    assert space.parameters["temperature"]["low"] == 0.0
    assert space.parameters["temperature"]["high"] == 1.0


def test_search_space_discrete():
    """Test discrete parameter definition."""
    space = SearchSpace()
    space.add_discrete("max_tokens", [128, 256, 512])

    assert "max_tokens" in space.parameters
    assert space.parameters["max_tokens"]["type"] == "discrete"
    assert space.parameters["max_tokens"]["values"] == [128, 256, 512]


def test_search_space_integer():
    """Test integer parameter definition."""
    space = SearchSpace()
    space.add_integer("n_examples", 1, 10)

    assert "n_examples" in space.parameters
    assert space.parameters["n_examples"]["type"] == "integer"
    assert space.parameters["n_examples"]["low"] == 1
    assert space.parameters["n_examples"]["high"] == 10


def test_search_space_categorical():
    """Test categorical parameter definition."""
    space = SearchSpace()
    space.add_categorical("model", ["gpt-4", "claude-3"])

    assert "model" in space.parameters
    assert space.parameters["model"]["type"] == "categorical"
    assert space.parameters["model"]["values"] == ["gpt-4", "claude-3"]


def test_search_space_sample():
    """Test sampling from search space."""
    space = SearchSpace()
    space.add_continuous("temperature", 0.0, 1.0)
    space.add_discrete("max_tokens", [128, 256, 512])
    space.add_categorical("model", ["gpt-4", "claude-3"])

    config = space.sample()

    assert "temperature" in config
    assert 0.0 <= config["temperature"] <= 1.0
    assert "max_tokens" in config
    assert config["max_tokens"] in [128, 256, 512]
    assert "model" in config
    assert config["model"] in ["gpt-4", "claude-3"]


def test_search_space_validate():
    """Test configuration validation."""
    space = SearchSpace()
    space.add_continuous("temperature", 0.0, 1.0)
    space.add_discrete("max_tokens", [128, 256, 512])

    # Valid config
    assert space.validate({"temperature": 0.5, "max_tokens": 256})

    # Invalid - out of range
    assert not space.validate({"temperature": 1.5, "max_tokens": 256})

    # Invalid - wrong discrete value
    assert not space.validate({"temperature": 0.5, "max_tokens": 1024})

    # Invalid - unknown parameter
    assert not space.validate({"temperature": 0.5, "unknown": 42})


# Test OptimizationResult


def test_optimization_result_creation():
    """Test optimization result creation."""
    from datetime import datetime

    start = datetime.now(UTC).isoformat()
    end = datetime.now(UTC).isoformat()

    result = OptimizationResult(
        best_config={"temperature": 0.5},
        best_score=0.95,
        history=[({"temperature": 0.3}, 0.85), ({"temperature": 0.5}, 0.95)],
        n_iterations=2,
        start_time=start,
        end_time=end,
    )

    assert result.best_config == {"temperature": 0.5}
    assert result.best_score == 0.95
    assert len(result.history) == 2
    assert result.n_iterations == 2


def test_optimization_result_improvement():
    """Test improvement calculation."""
    from datetime import datetime

    start = datetime.now(UTC).isoformat()
    end = datetime.now(UTC).isoformat()

    result = OptimizationResult(
        best_config={"temperature": 0.5},
        best_score=0.95,
        history=[({"temperature": 0.3}, 0.80), ({"temperature": 0.5}, 0.95)],
        n_iterations=2,
        start_time=start,
        end_time=end,
    )

    improvement = result.get_improvement()
    assert improvement > 0  # Positive improvement
    assert abs(improvement - 18.75) < 0.1  # (0.95 - 0.80) / 0.80 * 100 = 18.75%


def test_optimization_result_to_dict():
    """Test converting result to dictionary."""
    from datetime import datetime

    start = datetime.now(UTC).isoformat()
    end = datetime.now(UTC).isoformat()

    result = OptimizationResult(
        best_config={"temperature": 0.5},
        best_score=0.95,
        history=[({"temperature": 0.5}, 0.95)],
        n_iterations=1,
        start_time=start,
        end_time=end,
        metadata={"algorithm": "test"},
    )

    result_dict = result.to_dict()

    assert result_dict["best_config"] == {"temperature": 0.5}
    assert result_dict["best_score"] == 0.95
    assert result_dict["n_iterations"] == 1
    assert "duration_seconds" in result_dict
    assert result_dict["metadata"] == {"algorithm": "test"}


# Test RandomSearchOptimizer


@pytest.mark.asyncio
async def test_random_search_optimizer_basic():
    """Test basic random search optimization."""

    def agent_factory(config):
        return MockAgent(**config)

    search_space = {"temperature": (0.0, 1.0), "top_p": (0.0, 1.0)}

    optimizer = RandomSearchOptimizer(
        agent_factory=agent_factory, search_space=search_space, objective="accuracy", maximize=True
    )

    # Use expected values that match what MockAgent returns
    test_cases = [
        {"input": "What is 2+2?", "expected": "correct answer"},
        {"input": "What is the capital of France?", "expected": "correct answer"},
    ]

    result = await optimizer.optimize(test_cases, n_iterations=5)

    assert result.best_config is not None
    assert "temperature" in result.best_config
    assert "top_p" in result.best_config
    # With random search, at least one config should score > 0
    assert result.best_score >= 0
    assert len(result.history) == 5
    assert result.n_iterations == 5
    assert result.metadata["algorithm"] == "random_search"


@pytest.mark.asyncio
async def test_random_search_improvement():
    """Test that random search finds improvements."""

    def agent_factory(config):
        return MockAgent(**config)

    search_space = {"temperature": (0.0, 1.0), "top_p": (0.5, 1.0)}

    optimizer = RandomSearchOptimizer(
        agent_factory=agent_factory, search_space=search_space, objective="accuracy"
    )

    test_cases = [{"input": "Test", "expected": "correct answer"}] * 3

    result = await optimizer.optimize(test_cases, n_iterations=10)

    # With more iterations, should find reasonable config
    assert result.best_score > 0.3
    assert len(result.history) == 10


@pytest.mark.asyncio
async def test_random_search_discrete_space():
    """Test random search with discrete parameters."""

    def agent_factory(config):
        # Simulate that max_tokens=512 is best
        return MockAgent(temperature=config.get("temperature", 0.5))

    search_space = SearchSpace()
    search_space.add_continuous("temperature", 0.0, 1.0)
    search_space.add_discrete("max_tokens", [128, 256, 512])

    optimizer = RandomSearchOptimizer(
        agent_factory=agent_factory, search_space=search_space, objective="accuracy"
    )

    test_cases = [{"input": "Test", "expected": "correct answer"}] * 2

    result = await optimizer.optimize(test_cases, n_iterations=5)

    assert "max_tokens" in result.best_config
    assert result.best_config["max_tokens"] in [128, 256, 512]
